components/servo/generic/servo.go

```diff
@@ -0,0 +1,337 @@
+// Package servogeneric implements a generic servo
+package servogeneric
+
+import (
+	"context"
+	"math"
+	"time"
+
+	"github.com/edaniels/golog"
+	"github.com/pkg/errors"
+	viamutils "go.viam.com/utils"
+
+	"go.viam.com/rdk/components/board"
+	"go.viam.com/rdk/components/generic"
+	"go.viam.com/rdk/components/servo"
+	"go.viam.com/rdk/config"
+	"go.viam.com/rdk/operation"
+	"go.viam.com/rdk/registry"
+	"go.viam.com/rdk/utils"
+)
+
+const (
+	minDeg     float64 = 0.0
+	maxDeg     float64 = 180.0
+	minWidthUs uint    = 500  // absolute minimum width pwm width
+	maxWidthUs uint    = 2500 // absolute maximum width pwm width
+)
+
+type servoConfig struct {
+	Pin   string `json:"pin"`   // Pin a GPIO pin with pwm capabilities
+	Board string `json:"board"` // Board a board that exposes GPIO pins
+	// MinDeg minimum angle the servo can reach, note this doesn't affect PWM calculation
+	MinDeg *float64 `json:"min_angle_deg,omitempty"`
+	// MaxDeg maximum angle the servo can reach, note this doesn't affect PWM calculation
+	MaxDeg *float64 `json:"max_angle_deg,omitempty"`
+	// StartPos starting position of the servo in degree
+	StartPos *float64 `json:"starting_position_deg,omitempty"`
+	// Frequency when set the servo driver will attempt to change the GPIO pin's Frequency
+	Frequency *uint `json:"frequency_hz,omitempty"`
+	// Resolution resolution of the PWM driver (eg number of ticks for a full period) if left or 0
+	// the driver will attempt to estimate the resolution
+	Resolution *uint `json:"pwm_resolution"`
+	// MinWidthUS override the safe minimum width in us this affect PWM calculation
+	MinWidthUS *uint `json:"min_width_us"`
+	// MaxWidthUS Override the safe maximum width in us this affect PWM calculation
+	MaxWidthUS *uint `json:"max_width_us"`
+}
+
+// Validate ensures all parts of the config are valid.
+func (config *servoConfig) Validate(path string) ([]string, error) {
+	var deps []string
+	if config.Board == "" {
+		return nil, viamutils.NewConfigValidationFieldRequiredError(path, "board")
+	}
+	deps = append(deps, config.Board)
+	if config.Pin == "" {
+		return nil, viamutils.NewConfigValidationFieldRequiredError(path, "pin")
+	}
+	if config.StartPos != nil {
+		if config.MinDeg != nil && *config.StartPos < *config.MinDeg {
+			return nil, viamutils.NewConfigValidationError(path, errors.New("starting_position_degs cannot be lower than min_angle_deg"))
+		}
+		if config.MaxDeg != nil && *config.StartPos > *config.MaxDeg {
+			return nil, viamutils.NewConfigValidationError(path, errors.New("starting_position_degs cannot be higher than max_angle_deg"))
+		}
+		if *config.StartPos < minDeg || *config.StartPos > maxDeg {
```

Comment:
If I read this correctly, we do not support servos that can move more than 180 degrees. Is that right? I thought you could get some that could turn all the way around, potentially through multiple rotations. I kinda want to get rid of `minDeg` and `maxDeg` entirely and trust the user to specify `config.MinDeg` and `config.MaxDeg` themselves. ...and even that wouldn't work for servos that can turn arbitrary rotations.

Reply:
servo are supposed to be limited in range. For most of them the shaft is mounted on a potentiometer so the range can't go much further than 360, moreover they are only a finit number of position we can encode in a width of a pulse.
The accepted standard is 0-180, if we need a bigger range we would have to widen the microsecond delta

Reply:
That's a good example :) I don't think we need a separate model however, this servo also operates in the range of 500-2500 us so the principle of the code remains. We should loosen the requirements on the min/max deg. I will do that in a follow up PR when I come back since I don't think its a blocker right now